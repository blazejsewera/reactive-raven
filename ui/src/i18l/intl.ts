/**
 * Intl is an interface for an internationalization (i18l) tool
 * that can provide translations for page.
 */
export interface Intl {
  getTranslation: (text: string) => string
}

export const intl: Intl = { getTranslation: text => text }
