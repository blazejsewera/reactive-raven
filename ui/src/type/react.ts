import type { ReactElement, ReactNode } from "react";

type EmptyObject = Record<string, never>;

/**
 * Default children type in React. Every renderable type is included.
 */
export type DefaultChildren = ReactNode;

/**
 * No props, a.k.a. () => </> component
 */
export type EmptyProps = EmptyObject;

/**
 * No React Context
 */
export type EmptyContext = EmptyObject;

/**
 * Functional Component (strongly typed)
 */
export interface FC<P = EmptyProps, Ctx = EmptyContext> {
  (props: P, context?: Ctx): ReactElement | null;
  displayName?: string;
}
