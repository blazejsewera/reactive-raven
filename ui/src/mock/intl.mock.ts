import { Intl } from '../i18l/intl'

export const intlMock: Intl = { getTranslation: text => text }
