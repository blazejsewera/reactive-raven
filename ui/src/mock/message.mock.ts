import { Message } from '../type/message'
import { fiveSentenceLoremIpsum } from './asset/text/lipsum'
import githubIcon from './asset/icon/github-icon.svg'

const body = `This is a simple test message.

It spans over multiple lines.`

export const full: Message = {
  username: 'Primary',
  title: 'Test title',
  subtitle: '— primary',
  body,
  timestamp: '2 hours ago',
  id: '0',
}

export const fullWithImage: Message = {
  ...full,
  id: '1',
  userImgUri: githubIcon,
}

export const fullWithLoremIpsum: Message = {
  ...full,
  id: '2',
  body: fiveSentenceLoremIpsum,
}

export const partial: Message = {
  username: full.username,
  title: full.title,
  body: full.body,
  timestamp: full.timestamp,
  id: '3',
}

export const minimal: Message = {
  username: full.username,
  title: full.title,
  timestamp: full.timestamp,
  id: '4',
}

export const anotherUser: Message = {
  username: 'Other',
  title: 'Other title',
  subtitle: '— other',
  body,
  timestamp: '1 hour ago',
  id: '5',
}
