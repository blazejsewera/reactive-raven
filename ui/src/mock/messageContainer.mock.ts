import type { MessageContainerHandlers } from '../type/handler'

export const handlers: MessageContainerHandlers = {
  onClear: () => {},
}
