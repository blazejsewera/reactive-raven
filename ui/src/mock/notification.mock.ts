import type { Notification } from '../type/notification'
import { fiveSentenceLoremIpsum } from './asset/text/lipsum'
import githubIcon from './asset/icon/github-icon.svg'
import { NotificationCardHandlers } from '../type/handler'

const body = `#12 add some new amazing functionality

Closes #10. Changes both in 'core' and 'ui'. Needs additional work with this and that.`

const handlers: NotificationCardHandlers = {
  onArchive: () => {},
  onCheck: () => {},
  onSettings: () => {},
}

export const full: Notification = {
  appName: 'Github',
  title: 'New Pull Request',
  subtitle: 'â€” notipie',
  body,
  timestamp: '2 hours ago',
  id: '0',
  handlers,
}

export const fullWithImage: Notification = {
  ...full,
  id: '1',
  appImgUri: githubIcon,
}

export const fullWithLoremIpsum: Notification = {
  ...full,
  id: '2',
  body: fiveSentenceLoremIpsum,
}

export const partial: Notification = {
  appName: full.appName,
  title: full.title,
  body: full.body,
  timestamp: full.timestamp,
  id: '3',
  handlers,
}

export const minimal: Notification = {
  appName: full.appName,
  title: full.title,
  timestamp: full.timestamp,
  id: '4',
  handlers,
}
