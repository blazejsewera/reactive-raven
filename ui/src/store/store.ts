import { createStore } from 'redux'
import { Store } from 'redux'
import { Action } from './action/action'
import { Message } from '../type/message'
import { T_FAIL, T_REQUEST, T_SUCCESS } from './action/message/fetch'
import { T_PUSH, T_PUSH_ERROR } from './action/message/push'
import { merge } from '../util/message/merger'
import { T_DARKMODE_OFF, T_DARKMODE_ON, T_DARKMODE_TOGGLE } from './action/darkmode/set'
import { T_UPDATE_TIME } from './action/message/time'
import { updateTimeAll } from '../util/message/time'
import { config } from '../config/reactiveRaven'
import {
  T_FORM_CLEAR,
  T_FORM_UPDATE_BODY,
  T_FORM_UPDATE_SUBTITLE,
  T_FORM_UPDATE_TITLE,
  T_FORM_UPDATE_USERNAME,
} from './action/message/form'

export type State = {
  state: 'loading' | 'ok' | 'fail'
  messages: Message[]
  messageForm: Message
  isDarkMode: boolean
}

const emptyMessage: Message = {
  title: '',
  subtitle: '',
  body: '',
  username: '',
  timestamp: '',
}

const defaultState: State = {
  state: 'ok',
  messages: [],
  messageForm: emptyMessage,
  isDarkMode: false,
}

type Reducer = (state: State | undefined, action: Action) => State

const reducer: Reducer = (previousState = defaultState, action) => {
  switch (action.type) {
    case T_REQUEST:
      return { ...previousState, state: 'loading' }
    case T_SUCCESS:
      return { ...previousState, state: 'ok', messages: merge([...previousState.messages, ...action.messages]) } // PERF: possible room for optimization
    case T_FAIL:
      console.warn(action.message)
      return { ...previousState, state: 'fail' }
    case T_PUSH:
      return { ...previousState, state: 'ok', messages: merge([...previousState.messages, action.message]) } // PERF: possible room for optimization
    case T_PUSH_ERROR:
      console.warn(action.message)
      return { ...previousState, state: 'fail' }
    case T_UPDATE_TIME:
      return { ...previousState, messages: updateTimeAll(previousState.messages) }
    case T_FORM_UPDATE_TITLE:
      return { ...previousState, messageForm: { ...previousState.messageForm, title: action.title } }
    case T_FORM_UPDATE_SUBTITLE:
      return { ...previousState, messageForm: { ...previousState.messageForm, subtitle: action.subtitle } }
    case T_FORM_UPDATE_BODY:
      return { ...previousState, messageForm: { ...previousState.messageForm, body: action.body } }
    case T_FORM_UPDATE_USERNAME:
      return { ...previousState, messageForm: { ...previousState.messageForm, username: action.username } }
    case T_FORM_CLEAR:
      return { ...previousState, messageForm: emptyMessage }
    case T_DARKMODE_ON:
      return { ...previousState, isDarkMode: true }
    case T_DARKMODE_OFF:
      return { ...previousState, isDarkMode: false }
    case T_DARKMODE_TOGGLE:
      return { ...previousState, isDarkMode: !previousState.isDarkMode }
    default:
      return previousState
  }
}

const isDev = config.mode === 'dev'
const composeEnhancers = isDev
  ? (window as any).__REDUX_DEVTOOLS_EXTENSION__ && (window as any).__REDUX_DEVTOOLS_EXTENSION__()
  : undefined

export const store: Store<State, Action> = createStore(reducer, composeEnhancers)
export const { dispatch, subscribe } = store
