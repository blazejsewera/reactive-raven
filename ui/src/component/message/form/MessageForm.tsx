import * as React from 'react'
import { FC } from '../../../type/react'
import { Intl } from '../../../i18l/intl'
import { cx } from '../../../util/classname/cx'
import { IconAndControlsConnected as IconAndControls } from './section/iconAndControls/IconAndControls'
import { PaddingY } from './util/PaddingY'
import { dispatch, store } from '../../../store/store'
import {
  actionFormClear,
  actionFormUpdateBody,
  actionFormUpdateSubtitle,
  actionFormUpdateTitle,
  actionFormUpdateUsername,
} from '../../../store/action/message/form'
import { TitleConnected as Title } from './section/title/Title'
import { SubtitleConnected as Subtitle } from './section/subtitle/Subtitle'
import { BodyConnected as Body } from './section/body/Body'
import { UsernameConnected as Username } from './section/username/Username'
import { Message } from '../../../type/message'
import { pushMessage } from '../../../net/sync/request'
import { hash } from '../../../external/hash'

export interface MessageFormProps {
  intl: Intl
}

export const MessageForm: FC<MessageFormProps> = ({ intl }) => {
  return (
    <form
      className={cx(
        'rounded-3xl',
        'bg-gray-100',
        'dark:bg-gray-700',
        'w-128',
        'overflow-hidden',
        'shadow-lg',
        'relative',
      )}
    >
      <PaddingY />
      <Title {...{ intl, updateTitle }} />
      <Subtitle {...{ intl, updateSubtitle }} />
      <Body {...{ intl, updateBody }} />
      <Username {...{ intl, updateUsername }} />
      <IconAndControls intl={intl} onClear={clearForm} onSend={send} />
      <PaddingY />
    </form>
  )
}

const updateTitle = (title: string) => {
  dispatch(actionFormUpdateTitle(title))
}
const updateSubtitle = (subtitle: string) => {
  dispatch(actionFormUpdateSubtitle(subtitle))
}
const updateBody = (body: string) => {
  dispatch(actionFormUpdateBody(body))
}
const updateUsername = (username: string) => {
  dispatch(actionFormUpdateUsername(username))
}
const clearForm = () => {
  dispatch(actionFormClear())
}
const send = () => {
  const { messageForm } = store.getState()
  const message = addTimestamp(addId(messageForm))

  pushMessage(message)
  console.log('pushed')
}

const addId = (message: Message): Message => {
  const id = hash(JSON.stringify(message))
  return { ...message, id }
}

const addTimestamp = (message: Message): Message => {
  const now = new Date().toISOString()
  return { ...message, timestamp: now }
}
