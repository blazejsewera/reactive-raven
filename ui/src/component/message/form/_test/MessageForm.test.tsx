import * as React from 'react'
import { create as render } from 'react-test-renderer'
import { MessageForm } from '../MessageForm'
import { intlMock } from '../../../../mock/intl.mock'
import { Provider } from 'react-redux'
import { store } from '../../../../store/store'

describe('MessageForm component', () => {
  it('renders correctly', () => {
    // given
    const intl = intlMock
    const tree = render(
      <Provider store={store}>
        <MessageForm intl={intl} />
      </Provider>,
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
