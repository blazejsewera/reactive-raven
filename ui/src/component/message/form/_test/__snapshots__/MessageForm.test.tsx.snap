// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageForm component renders correctly 1`] = `
<form
  className="rounded-3xl bg-gray-100 dark:bg-gray-700 w-128 overflow-hidden shadow-lg relative"
>
  <div
    className="flex flex-row w-full h-2"
  >
    <div
      className="basis-1/4 bg-gray-300 dark:bg-gray-800"
    />
    <div
      className="basis-3/4 bg-gray-100 dark:bg-gray-700"
    />
  </div>
  <label
    className="flex flex-row w-full"
    htmlFor="form-title"
  >
    <div
      className="bg-gray-300 text-gray-800 dark:bg-gray-800 dark:text-gray-200 flex shrink-0 font-bold text-left w-full sm:text-right sm:basis-1/4 p-2"
    >
      <span
        className="my-auto w-full"
      >
        title
      </span>
    </div>
    <div
      className="w-full m-2 mr-4 basis-3/4 flex"
    >
      <input
        className="outline-none active:outline-none focus:outline-none ring-0 active:ring-0 focus:ring-0 bg-white text-gray-800 border-gray-300 active:border-gray-500 focus:border-gray-400 dark:bg-gray-800 dark:text-gray-200 dark:border-gray-600 dark:active:border-gray-400 dark:focus:border-gray-500 w-full rounded-md border-2 text-xl font-bold w-full p-1"
        id="form-title"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
  </label>
  <label
    className="flex flex-row w-full"
    htmlFor="form-subtitle"
  >
    <div
      className="bg-gray-300 text-gray-800 dark:bg-gray-800 dark:text-gray-200 flex shrink-0 font-bold text-left w-full sm:text-right sm:basis-1/4 p-2"
    >
      <span
        className="my-auto w-full"
      >
        subtitle
      </span>
    </div>
    <div
      className="w-full m-2 mr-4 basis-3/4 flex"
    >
      <input
        className="outline-none active:outline-none focus:outline-none ring-0 active:ring-0 focus:ring-0 bg-white text-gray-800 border-gray-300 active:border-gray-500 focus:border-gray-400 dark:bg-gray-800 dark:text-gray-200 dark:border-gray-600 dark:active:border-gray-400 dark:focus:border-gray-500 w-full rounded-md border-2  w-full p-1"
        id="form-subtitle"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
  </label>
  <label
    className="flex flex-row w-full"
    htmlFor="form-body"
  >
    <div
      className="bg-gray-300 text-gray-800 dark:bg-gray-800 dark:text-gray-200 flex shrink-0 font-bold text-left w-full sm:text-right sm:basis-1/4 p-2"
    >
      <span
        className="mb-auto mt-1 w-full"
      >
        body
      </span>
    </div>
    <div
      className="w-full m-2 mr-4 basis-3/4 flex"
    >
      <textarea
        className="outline-none active:outline-none focus:outline-none ring-0 active:ring-0 focus:ring-0 bg-white text-gray-800 border-gray-300 active:border-gray-500 focus:border-gray-400 dark:bg-gray-800 dark:text-gray-200 dark:border-gray-600 dark:active:border-gray-400 dark:focus:border-gray-500 w-full rounded-md border-2 w-full p-1 min-h-9 max-h-48"
        id="form-body"
        onChange={[Function]}
        value=""
      />
    </div>
  </label>
  <label
    className="flex flex-row w-full"
    htmlFor="form-username"
  >
    <div
      className="bg-gray-300 text-gray-800 dark:bg-gray-800 dark:text-gray-200 flex shrink-0 font-bold text-left w-full sm:text-right sm:basis-1/4 p-2"
    >
      <span
        className="my-auto w-full"
      >
        username
      </span>
    </div>
    <div
      className="w-full m-2 mr-4 basis-3/4 flex"
    >
      <input
        className="outline-none active:outline-none focus:outline-none ring-0 active:ring-0 focus:ring-0 bg-white text-gray-800 border-gray-300 active:border-gray-500 focus:border-gray-400 dark:bg-gray-800 dark:text-gray-200 dark:border-gray-600 dark:active:border-gray-400 dark:focus:border-gray-500 w-full rounded-md border-2  w-full p-1"
        id="form-username"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
  </label>
  <div
    className="w-full flex flex-row"
  >
    <div
      className="bg-gray-300 text-gray-800 dark:bg-gray-800 dark:text-gray-200 flex shrink-0 font-bold text-left w-full sm:text-right sm:basis-1/4 p-2"
    >
      <span
        className="my-auto w-full"
      >
        icon
      </span>
    </div>
    <div
      className="w-full m-2 mr-4 basis-3/4 flex"
    >
      <div
        className="my-auto h-12 w-12 rounded-full overflow-clip overflow-hidden inline-block cursor-default"
      >
        <svg
          viewBox="0 0 1 1"
        >
          <rect
            fill="#111827"
            height={1}
            width={1}
          />
          <text
            dy="0.178"
            fill="#FFF"
            style={
              Object {
                "fontFamily": "Inter",
                "fontSize": 0.5,
                "fontWeight": "bold",
              }
            }
            textAnchor="middle"
            x={0.5}
            y={0.5}
          >
            
          </text>
        </svg>
      </div>
      <div
        className="flex flex-row h-full ml-auto"
      >
        <button
          className="my-auto"
          onClick={[Function]}
          type="button"
        >
          <svg
            className="text-gray-500 dark:text-gray-400"
            fill="none"
            height={24}
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
            viewBox="0 0 24 24"
            width={24}
            xmlns="http://www.w3.org/2000/svg"
          >
            <polyline
              points="3 6 5 6 21 6"
            />
            <path
              d="M19 6v14a2 2 0 0 1-2 2H7a2 2 0 0 1-2-2V6m3 0V4a2 2 0 0 1 2-2h4a2 2 0 0 1 2 2v2"
            />
            <line
              x1="10"
              x2="10"
              y1="11"
              y2="17"
            />
            <line
              x1="14"
              x2="14"
              y1="11"
              y2="17"
            />
          </svg>
        </button>
        <button
          className="my-auto ml-4 flex w-14 h-14 rounded-full bg-gray-700 dark:bg-gray-400"
          onClick={[Function]}
          type="submit"
        >
          <svg
            className="text-gray-100 dark:text-gray-600 m-auto -translate-x-px translate-y-px"
            fill="none"
            height={24}
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
            viewBox="0 0 24 24"
            width={24}
            xmlns="http://www.w3.org/2000/svg"
          >
            <line
              x1="22"
              x2="11"
              y1="2"
              y2="13"
            />
            <polygon
              points="22 2 15 22 11 13 2 9 22 2"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
  <div
    className="flex flex-row w-full h-2"
  >
    <div
      className="basis-1/4 bg-gray-300 dark:bg-gray-800"
    />
    <div
      className="basis-3/4 bg-gray-100 dark:bg-gray-700"
    />
  </div>
</form>
`;
